File: Merge Sort/merge_sort.py
def merge_sort(arr):
    # Если длина массива 0 или 1, он уже отсортирован
    if len(arr) > 1:
        mid = len(arr) // 2
        left_part = arr[:mid]
        right_part = arr[mid:]

        # Рекурсивно сортируем левую и правую часть
        merge_sort(left_part)
        merge_sort(right_part)

        # Индексы для итерирования
        i = j = k = 0

        # Слияние отсортированных частей
        while i < len(left_part) and j < len(right_part):
            if left_part[i] < right_part[j]:
                arr[k] = left_part[i]
                i += 1
            else:
                arr[k] = right_part[j]
                j += 1
            k += 1

        # Добавляем оставшиеся элементы левой части (если есть)
        while i < len(left_part):
            arr[k] = left_part[i]
            i += 1
            k += 1

        # Добавляем оставшиеся элементы правой части (если есть)
        while j < len(right_part):
            arr[k] = right_part[j]
            j += 1
            k += 1

# Пример использования
numbers = [38, 27, 43, 3, 9, 82, 10]
merge_sort(numbers)
print(numbers)  # Результат: [3, 9, 10, 27, 38, 43, 82]



"""
В коде есть несколько ключевых моментов:

Рекурсия: если массив достаточно большой, разделяем его на две части, сортируем каждую.
Фаза слияния (while i < len(left_part) and j < len(right_part)): здесь мы сравниваем первые неиспользованные элементы из левой и правой части и помещаем меньший из них в итоговый массив.
Добавление «хвоста»: когда закончились элементы в одной из частей, нужно просто добавить остаток другой части.
"""

File: Merge Sort/merge_sort.js
function mergeSort(arr) {
  // Базовый случай: массив из одного элемента уже отсортирован
  if (arr.length <= 1) return arr;
  
  // Делим массив пополам
  const mid = Math.floor(arr.length / 2);
  const left = arr.slice(0, mid);
  const right = arr.slice(mid);
  
  // Рекурсивно сортируем обе половины
  return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
  const result = [];
  let i = 0;
  let j = 0;
  
  // Сравниваем элементы из обеих половин и добавляем меньший
  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]) {
      result.push(left[i]);
      i++;
    } else {
      result.push(right[j]);
      j++;
    }
  }
  
  // Добавляем оставшиеся элементы
  result.push(...left.slice(i));
  result.push(...right.slice(j));
  
  return result;
}

// Пример использования
console.log(mergeSort([64, 34, 25, 12, 22, 11, 90]));
// Результат: [11, 12, 22, 25, 34, 64, 90]
